### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by signalwerk** [fontra/fontra#194] [General]
I saw this [tweet](https://twitter.com/blackfoundry/status/1595744733770706944) today that you’re looking for a web developer. I applaud this decision and also your effort! This is a great project! 

## tl;dr
> I would like to know if you have a roadmap for this project and if you have more architectural decisions/documentation to share here. ((I already saw `/docs`))

## Backstory
When I saw the [video of the fontra announcement](https://www.youtube.com/watch?v=Xhw5iCTK1XI) a few weeks back I was super excited, but I just had a brief look at the code and my first reaction was: love the idea, love the effort, but the current state was not only excitement (who am I to judge anyway...).  
After the tweet today, I thought: the shape of the next generation font-editor should be nothing but awesome. So, I installed today the code and had a look at it. 

## What I really like
* Many details that are essential for font-designers are already now implemented (shortcuts, smooth zooming, …)
* Rendering done in canvas
* Love the approach of the change-recorder and the communication of state via websockets
* Already a significant amount of code 

## What I was concerning
* I don’t see a sustainable financial model right now to the project. Are there plans for it? I wouldn't like to see it ending like [metapolator](https://github.com/metapolator/metapolator).
* GPL-3 license (JS-World might work different)
* No modern front-end framework (react/vue/svelte/…) and no modern state manager (store/mutation concept)
* Many things are built from scratch (only two external libs are used for testing) – I guess that’s the same discussion [as between Runebender and MFEK](https://github.com/MFEK/docs/blob/master/doc/Why.md)

## Thank you!
Please do not take my comment here in any form as an attack or rant. I respect this project and the people behind it a lot!

**Comment by signalwerk** [comment]
Thank you, @justvanrossum for the [road-map](https://github.com/BlackFoundryCom/fontra/pull/205/commits/aeacaeca18cc65aae46ec2d632fbc27336166a26) you provided in the meanwhile.

**Comment by signalwerk** [comment]
Thank you, @davelab6 for your [commit](https://github.com/BlackFoundryCom/fontra/commit/f18fc6677046c7a6b31e813efddbebd9d988f2d7). I guess google is therefore supporting the project. That's wonderful to hear!

**Comment by davelab6** [comment]
@signalwerk sorry my delay. I am only peripherally involved in Fontra, but I am funding development and setting high level strategic goals, so I am taking the liberty to reply :)

> I don’t see a sustainable financial model right now to the project. Are there plans for it? I wouldn't like to see it ending like [metapolator](https://github.com/metapolator/metapolator).

No, the financial model is the same as metapolator. In some ways this is metapolator 2. 

> GPL-3 license (JS-World might work different)

You'll need to be more specific, but basically I would say, almost all libre font editors are/were GPL, and many web **applications** are GPL, while I agree that many web **libraries** do not have copyleft protections for users.

> No modern front-end framework (react/vue/svelte/…) and no modern state manager (store/mutation concept)
>
> Many things are built from scratch (only two external libs are used for testing) – I guess that’s the same discussion [as between Runebender and MFEK](https://github.com/MFEK/docs/blob/master/doc/Why.md)

I'll answer those together: Generally I am quite against front end frameworks these days as I believe they rot too fast; but again I need you to be more specific and explain exactly what problem fontra has that you have experienced (as a user, or as a developer) tha ta framework or modern state manager would actually solve.

**Comment by signalwerk** [comment]
Thank you @davelab6 for your Answer. I think my post was mainly a reaction/request to the [tweet](https://twitter.com/blackfoundry/status/1595744733770706944) and for me to understand if it would be worth considering it. It really sounds like a dream job, and the project itself is just wonderful! But since we are half a year later, I hope you found someone for hire and therefore I'm going to close the comment. You basically answered my questions, and I guess I just have another perspective on it. If the project dies as metapolator died, then it doesn't matter what the roadmap/licence/technology is. Then it's driven by some people that see it as a paid job, and afterwards there is no community left. 
But we didn't manage to get an opensource font-editor that has a wide (dev-)community until now and I think this is really a bummer, and I am trying to understand why it is like that. 

I can't speak for the community, but I can speak for myself why fontra is not attracting me a lot at the moment:

1. **History** → If the setup for fontra is not different from metapolator it will die.
2. **Licence** → Sure I love open source. Sure I'm aware of the copyleft-arguments. In my personal (JS/TS) world, I just saw many clients choosing projects and support (in the form of work) based on Licence question. And while copyleft is nice, it doesn't matter if nobody is interested in contributing anyway. If someone want's to fork their private copy of fontra; I would not hold that person back. I know over the long run that person/company is an ally.
3. **Technology** → fontra is starting from scratch. doing everything fresh and without a framework sounds nice, but it will also be a lot of work. Plus for me, as a developer, I would like to be quickly onboarded to a project. Standards help for that. Standards and safety guards. For that, frameworks, linters and type safety helps. Sure, you can do your own state-manager, and you can start your own state-to-gui-logic because you fear of framework fatigue. But to be honest, I also have a `innerHTML` fatigue 🫣. There is a reason why corporations use frameworks and the reason is most of the time; business value. For me, that's not too different from community value.

I will give you an example (doesn't matter whether it's realistic or not). Imagine you need Boolean-Operations in the Frontend-Code. This is really a hard problem and I saw only one robust and open implement in JavaScript so far. So, what do you do now? You start to implement it yourself. Sure. You can do a good Implementation in 2–3 days. But this will not lead to a robust implementation. You spend more time, more tests, more work. And in the end, you have spent a lot of time and energy on boolean operations. A very common problem, but now nobody will take over your implementations. Why? You have the wrong licence (non GPL-3 can't use it), you have it in your project and not separate, you didn't consider other cases in your first implementation (colours, outlines, ...). You included complexity in your project, and now you need to maintain it. You're doing stuff like that right now, but maybe it's not that obvious. If you don't use external packages you basically say; I have a unique case, and I'm willing to maintain all the complexity myself.

But again; I completely respect your decisions and I love the project. And at one point the project might be so strong and advanced that all I said here sounds stupid, and I will happily use it again. Looking forward to that moment!

Go Team Fontra!

**Reply by davelab6** [reply]
> I hope you found someone for hire and therefore I'm going to close the comment.

I believe the position is filled but its a question for @JeremieHornus :) 

> If the project dies as metapolator died, then it doesn't matter what the roadmap/licence/technology is. Then it's driven by some people that see it as a paid job, and afterwards there is no community left.

Well, I don't really see a difference between paid-jobs and "community." Writing a complete editor is large-scale and intensive work, and intensive work requires basic social reproduction (rent, food, healthcare, etc) that has to be paid somehow. The problem with metapolator was that it in the funded period it did not develop sufficient functionality to become actually useful to anyone for any commercial purpose, such that their own usage activity neither created a self reinforcing feedback loop of self-interested development ("Today while using the product I encountered this bug and I fixed it, here's a PR") nor demand in the marketplace for development ("Today while using the product for a paid project I encountered this bug and I am seeking someone to fix it, here's a coin"). In both cases, there is public development, larger commercial interests involved, and people paid to work on the editor. 

It might be possible that Fontra will reach such "lift-off velocity" before "running out of runway", since Black Foundry is using it for one project they have contracted with Google for, and once that project is delivered, they may have other similar projects from other clients, and continue to fund development that way. Google's CJK project is not their first CJK branding project, after all, and if the Google one is seen as successful, it may be a strong case study for them to market themselves. 

And also, its possible that Just and Jeremie will make different decisions to me, Simon, Lasse, Peter and the others involved in directing Metapolator, such that even with all else being equal, they actually do product something useful in day to day workflows before their time runs out. Having a specific typeface deliverable will help focus, I think.

> But we didn't manage to get an opensource font-editor that has a wide (dev-)community until now and I think this is really a bummer, and I am trying to understand why it is like that.

Few people want to write the whole thing, and are content enough with someone else's foundation and their scripting API, as long as it offers sufficient functionality to get their immediate needs me. Only either 'true believers' of the "libre software user freedom movement" find this ethically unacceptable - e.g. [MFEK.org](https://mfek.org/) - or people with a LOT of subsidized time, retirees or students in FontForge and TruFont - are willing to write a whole thing. The rest (glyphr, runebender) are personal projects to use a font editor as a vehicle to learn and explore something else (web app development, rust GUIs), and as hobby horses are subject to whims and winds. 

> History → If the setup for fontra is not different from metapolator it will die.

Perhaps it is different enough, as above.

> Licence → Sure I love open source. Sure I'm aware of the copyleft-arguments. In my personal (JS/TS) world, I just saw many clients choosing projects and support (in the form of work) based on Licence question. And while copyleft is nice, it doesn't matter if nobody is interested in contributing anyway. If someone want's to fork their private copy of fontra; I would not hold that person back. I know over the long run that person/company is an ally.

Sure, I know for example Microsoft folks won't contribute. Maybe Adobe also may have some concerns. But GPL allows private forks; its proprietary forks that are prevented, and it seems unlikely to me that MS or Adobe would want to offer one. 

Perhaps you are right and an ally will silently approach and flee without any discussion, but my experience with Font Bakery (which had at some point a GPL dependency) is that in fact if they are bringing resources to the table and license is the only concern, they will raise it and we will discuss it.

> Technology → fontra is starting from scratch. doing everything fresh and without a framework sounds nice, but it will also be a lot of work. Plus for me, as a developer, I would like to be quickly onboarded to a project. Standards help for that. Standards and safety guards. For that, frameworks, linters and type safety helps. Sure, you can do your own state-manager, and you can start your own state-to-gui-logic because you fear of framework fatigue. But to be honest, I also have a innerHTML fatigue 🫣. There is a reason why corporations use frameworks and the reason is most of the time; business value. For me, that's not too different from community value.

I don't find anything to disagree with in this, however, at this stage, Fontra is a "sketch" phase for how it should be - Just is exploring what can work, and isn't actually seeking any contributions (AFAIK.) As the project matures into another phase, declaring (and setting up linters or other automation) for standards is certainly a good idea for me. Type safety checking without requiring a rewrite to typescript is good in a phase of performance optimization, but in sketching, seems to me that it is helpful and productive to allow duck typing.

> I will give you an example (doesn't matter whether it's realistic or not). Imagine you need Boolean-Operations in the Frontend-Code. This is really a hard problem and I saw only one robust and open implement in JavaScript so far. So, what do you do now? You start to implement it yourself. Sure. 

We'll see, but I would suspect that for this specific example it would be good to consider kurbo's wasm, which has more and more [b-ops functions appearing](https://raphlinus.github.io/curves/2023/04/18/bezpath-simplify.html). It is also trivial to take some functionality out into its own repo with its own license, once out of the sketch phrase. 

> You're doing stuff like that right now

I'm curious if you can name the specific cases :) 

> But again; I completely respect your decisions and I love the project. And at one point the project might be so strong and advanced that all I said here sounds stupid, and I will happily use it again. Looking forward to that moment!
>
> Go Team Fontra!

Appreciated! We are all in Paris this week for ATypI, BTW

**Comment by signalwerk** [comment]
@davelab6 thank you for your extensive answer and caring so much about Open Source in the field and font production.

> I don't really see a difference between paid-jobs and "community."

I think the main difference is; if you build on top of a community, you need to have enthusiasts and people who care. That's different to paying someone. You can have a kick-start with one-time-money, but later you need a sustainable model. Ideally, with people who care and earn money. That's probably the ideal solution, and I hope fontra can reach that level before running out of google-funding.

> But GPL …

I guess that's a cultural and historical discussion. I respect all the GPL-Believes, and it brought us many good. I just have a different understanding nowadays – but that's most likely only a matter of preference and not worth an additional argument. We can read the pros and cons in every libre-project.

> … Fontra is a "sketch" phase for how it should be - Just is exploring what can work, and isn't actually seeking any contributions (AFAIK.) 

It's a good sketch! I might have read it just as more than a sketch. But definitely you can go wilder if you see it as a sketch. 

>  It is also trivial to take some functionality out into its own repo with its own license, …

You're right. Ideally, it would be cool not only to break out certain parts, but also to bring in other parts that already have a community.

> I'm curious if you can name the specific cases :)

It's basically the whole code-base that is stating it's an individual case nobody had before. Or as you named it, and that's probably also true here: it's a sketch.  
If you would like to have one example named, I can pick one: `/src/fontra/client/core/transform.js` basically another 2d-matrix-lib. 

>  We are all in Paris this week for ATypI

I'm based in Switzerland and usually can't attend to ATypI. But let's hope a [Libre Graphics Meeting](https://libregraphicsmeeting.org/) will be our next chance to meet.

**Reply by justvanrossum** [reply]
FWIW, `transform.js` is a partial port of fonttools' `transform.py`, which is used on Fontra's Python side. For our purposes it is very useful to have a 1-on-1 translation. Using an arbitrary other 2d matrix lib (or even the best/most popular/whatever one on npm) would likely cause more friction for us than this tiny rewrite.

**Reply by davelab6** [reply]
Thanks @justvanrossum - that's an excellent reason for transform.js to exist, I think - and your first commit (per https://github.com/googlefonts/fontra/blame/main/src/fontra/client/core/transform.js) did explain this on line 1, albeit in a terse way. I've made PR #506 to pull this fragment of lore into that comment line :)

**Comment by BlackFoundry** [comment]
Fontra is a WIP tool, that's publicly open for anyone interested to try out. I wouldn't state it's a sketch.
At Black Foundry we have plans for using it, and we are already using it, and more and more as new features are added.
Our focus and priorities are probably not what most type designer expect, but base on what _we_ need first for a specific large scale CJK project.
We are also planing to use it for Latin typeface design and more… so if you are interested, bear with us. But our goal is also to make it Black Foundry's main font editor.
It will survive as long as the foundry does if we manage to make it work for our own purposes :)

**Reply by davelab6** [reply]
> I wouldn't state it's a sketch

Then it has already matured from the last time I discussed the development strategy with @justvanrossum as I'm pretty sure that is how he explained it at that time :)

> It will survive as long as the foundry does if we manage to make it work for our own purposes :)

That would be excellent - although I guess it remains to be seen if you will continue _public_ development – and I guess that depends on our ability to grow a user-to-developer/contributor community, such that private development by BlackFoundry wouldn't make sense because public development by the rest of the community would continue, and it would become expensive to keep up.

**Comment by signalwerk** [comment]
@justvanrossum @BlackFoundry thank you for your answers and further explanations. I don't want to rock the boat at all. My first post was in response to get a feeling if I should apply to the job-offer and for that, I only had the code to build my opinion and wanted to get more insides. Sorry if I seam to be ungrateful. I'm very excited about fontra and I appreciate all your time, money and sweat you're investing in it. I wish the project only the best.

**Reply by justvanrossum** [reply]
> if I should apply to the job-offer

You could have just said so :) If you're still vaguely interested, please send your credentials to hello@black-foundry.com and perhaps we can talk.

**Reply by signalwerk** [reply]
> You could have just said so :)

I thought that was clear from my very first sentence in that whole thread (6 months ago), since I reacted to the tweet when you were looking for someone. But I see… I made myself not clear enough. 🤓

**Comment by davelab6** [comment]
> @davelab6 thank you for your extensive answer and caring so much about Open Source in the field and font production.

You are very welcome :)
 
> > I don't really see a difference between paid-jobs and "community."
> 
> I think the main difference is; if you build on top of a community, you need to have enthusiasts and people who care.

Perhaps, although I suspect that since @justvanrossum has designed Fontra with several plugin APIs, then what we will see is that the majority of the "people who care" user-developers' code is written to those APIs, just the same as with RoboFont, GlyphsApp, and Fontlab, rather than working on internals; and if any of those 3 editors was to turn all their private code public under a libre license, it would suffer the same problem: Most people enjoying writing GlyphsApp extensions are not sufficiently fluent with Objective C that they could go out and successfully apply for a salary job as a software engineer at some mac app shop. 

> That's different to paying someone. 

I agree, in that, when things are made for sale, rather than for direct use, there is always shift in priorities. Sometimes this is a large and kind of grotesque degradation of the nature of the thing, where the power of marketing/sales/position compels us to use the things even though as we spend time with it we realize it is junk, that it sells well but most customers are grumpy, disillusioned, and (sooner or later) it has high churn and no long-term loyal customers or future. Sometimes this is very subtle, and the company enjoys a strong and happy community around it that is just as good as the best libre software projects. Surely we all know of proprietary applications in the font editor space and elsewhere we have gone through phases of these situations.... and since I work with 100s of type designers, I am always amused when I hear the diametric opposite opinions about the same editor product within a short period :) 

But, conversely, when things are made purely for immediate direct use, and no attention is paid for 3rd party users, then it also becomes a different kind of grotesque degradation of the nature of the thing... users who show up expecting the developers to care about them are just as grumpy and churn away disillusioned, because the developer does not care at all about them. No documentation? Read the source code! No basic feature I expect? Make it and send a PR! Can I pay you to add the thing I want? No, money and libre software should not mix! (Or worse :) 

Ultimately my point here is that "cultivating a community of enthusiastic people who care" is 100% orthogonal to the use of money to advance the project. 

> You can have a kick-start with one-time-money, but later you need a sustainable model. Ideally, with people who care and earn money. That's probably the ideal solution, and I hope fontra can reach that level before running out of google-funding.

I agree that the ideal solution will be adoption by enough established foundries that they are self-interested to continue development publicly, and BlackFoundry is adopter #1. I hope we'll find a #2 before 2024 :)

> > But GPL …
> 
> I guess that's a cultural and historical discussion. I respect all the GPL-Believes, and it brought us many good. I just have a different understanding nowadays – but that's most likely only a matter of preference and not worth an additional argument. We can read the pros and cons in every libre-project.

:)

> > … Fontra is a "sketch" phase for how it should be - Just is exploring what can work, and isn't actually seeking any contributions (AFAIK.) 
> 
> It's a good sketch! I might have read it just as more than a sketch. But definitely you can go wilder if you see it as a sketch. 

It seems it already matured beyond that stage, and so I recommend to the team to start thinking about carving things out - but this is probably also not a priority anytime soon, unless there is a direct request for it to respond to.

> >  It is also trivial to take some functionality out into its own repo with its own license, …
> 
> You're right. Ideally, it would be cool not only to break out certain parts, but also to bring in other parts that already have a community.

I would expect harfbuzz to be an early such part to bring in :) 

> > I'm curious if you can name the specific cases :)
> 
> It's basically the whole code-base that is stating it's an individual case nobody had before. Or as you named it, and that's probably also true here: it's a sketch.  
> If you would like to have one example named, I can pick one: `/src/fontra/client/core/transform.js` basically another 2d-matrix-lib. 

Any other examples? :) 

> >  We are all in Paris this week for ATypI
> 
> I'm based in Switzerland and usually can't attend to ATypI. But let's hope a [Libre Graphics Meeting](https://libregraphicsmeeting.org/) will be our next chance to meet.

I hope there will be another successful LGM soon, but it seems the community has dispersed back into the individual projects; the 2021 event (now 25 months ago) had very few proposals (so few no committee was needed to triage them and they were all accepted) and no one has stepped up in 2021 or 2022 to run an event since then.... 

The GIMP community seems smaller than ~10 years ago, and is mainly a project of 2 people now, per https://www.gimp.org/news/2023/01/29/2022-annual-report - who appear to receive $1-1.5k USD/month on Patreon each, which keeps them going. If $50k year is "sustainable" for GIMP, I hope that it will be sustainable for Fontra too :) (Then again, the [GIMP bitcoin wallet](https://www.blockchain.com/explorer/addresses/btc/1NVMCeoBfTAJQ1qwX2Dx1C8zkcRCQWwHBq) is currently worth over $500k, and was worth over $1M a couple years ago.... so maybe money is not the limiting factor there :)
